src/main/scala/org/kframework/kore/interfaces/Interfaces.scala

```diff
@@ -481,61 +481,183 @@ object pattern {
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Sentence
+  }
+
+  trait Import extends Sentence {
+    val name: p.Name
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Import
+  }
+
+  object Import {
+    def unapply(arg: Import): Option[(p.Name, Attributes)] = Some(arg.name, arg.att)
+  }
+
+  trait SortDeclaration extends Sentence {
+    val sort: p.Sort
+
+    override lazy val sorts = Set(sort)
+
+    override def onAttributes(f: p.Pattern => p.Pattern): SortDeclaration
+  }
+
+  object SortDeclaration {
+    def unapply(arg: SortDeclaration): Option[(p.Sort, Attributes)] = Some(arg.sort, arg.att)
+  }
+
+  trait SymbolDeclaration extends Sentence {
+    val sort: p.Sort
+    val symbol: p.Symbol
+    val args: Seq[p.Sort]
+
+    // TODO: Should `sorts` include the `args` here as well?
+    override lazy val sorts = Set(sort)
+    override lazy val symbols = Set(symbol)
+
+    override def onAttributes(f: p.Pattern => p.Pattern): SymbolDeclaration
+  }
+
+  object SymbolDeclaration {
+    def unapply(arg: SymbolDeclaration): Option[(p.Sort, p.Symbol, Seq[p.Sort], Attributes)] = Some(arg.sort, arg.symbol, arg.args, arg.att)
+  }
+
+  trait Rule extends Sentence {
+    val pattern: p.Pattern
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Rule
+  }
+
+  object Rule {
+    def unapply(arg: Rule): Option[(p.Pattern, Attributes)] = Some(arg.pattern, arg.att)
+  }
+
+  trait Axiom extends Sentence {
+    val pattern: p.Pattern
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Axiom
+  }
+
+  object Axiom {
+    def unapply(arg: Axiom): Option[(p.Pattern, Attributes)] = Some(arg.pattern, arg.att)
   }
 
 }
 
+object builders {
+
+  trait OuterBuilders {
+    import outer._
+    import pattern.{Name, Sort, Symbol, Pattern}
+
+    def Definition(modules: Seq[Module], att: Attributes): Definition
+    def Module(name: Name, sentences: Seq[Sentence], att: Attributes): Module
+    def Import(name: Name, att: Attributes): Import
+    def SortDeclaration(sort: Sort, att: Attributes): SortDeclaration
+    def SymbolDeclaration(sort: Sort, symbol: Symbol, args: Seq[Sort], att: Attributes): SymbolDeclaration
+    def Rule(pattern: Pattern, att: Attributes): Rule
+    def Axiom(pattern: Pattern, att: Attributes): Axiom
+  }
+
+  trait PatternBuilders {
+    import pattern._
+
+    def Variable(_1: Name, _2: Sort): Variable
+    def DomainValue(_1: Symbol, _2: Value): DomainValue
+    def Top(): Top
+    def Bottom(): Bottom
+    def Not(_1: Pattern): Not
+    def Next(_1: Pattern): Next
+    def And(_1: Pattern, _2: Pattern): And
+    def Or(_1: Pattern, _2: Pattern): Or
+    def Implies(_1: Pattern, _2: Pattern): Implies
+    def Equals(_1: Pattern, _2: Pattern): Equals
+    def Exists(_1: Variable, _2: Pattern): Exists
+    def ForAll(_1: Variable, _2: Pattern): ForAll
+    def Rewrite(_1: Pattern, _2: Pattern): Rewrite
+    def Application(_1: Symbol, args: Seq[Pattern]): Application
+  }
+
```

Comment:
This is related to what Daejun was arguing about in a recent K meeting, and I said we can make it work. Yes, we can make it work, but this separation still exists, and I think we should leave in the interfaces.

Some backends (e.g., both the Java one and Kale) can only construct correct pattern objects once they have seen the entire definition -- this is natural because some symbols only make sense once several sentences have been seen, e.g., you can only construct an assoc once you can see the symbol for its identity element. 

E.g.,
```
val kale = new KaleBuilders()
val setPatternDeclaredVeryEarly = kale.Application(Symbol("_ _"), Seq(1, 2)) // 1 and 2 don't matter
val setOpSentence = kale.SymbolDeclaration(Exp, Symbol("_ _"), ...)
val setIdentity = kale.SymbolDeclaration(Exp, Symbol(".Set"), ...)
// and so on until eventually we get to
val definition = kale.Definition(...)
// now we take the definition and try to rewrite with it
definition.step(setPatternDeclaredVeryEarly)
```

The problem is that that Kale only knows how to properly store the set once it sees the entire definition, which is after the set pattern was created. So now it would have to store the Set in a different, "default", format convert to the right implementation every time, which defeats the whole purpose of having a pattern interface. Well, once the definition is initialized, Kale could make the smarter builders available. Right, but should we expose them? Should it be in the monolithic `KaleBuilders`? In this case, `KaleBuilders` will only be able to implement Patterns for one definition and be done, because otherwise it will become ambiguous which Pattern type it constructs. And it would make the Builders a mutable object that cannot be shared between multiple threads example. And with more complex behavior. Essentially, I strongly believe we should not go on that route.

If we keep the interfaces separate, we can do the following:
```
// we could have, for convenience:
trait Builders extends OuterBuilders with PatternBuilders

// but... 
trait DefaultPatternBuilders extends PatternBuilders { ... }
class KaleBuilders extends Builders with DefaultPatternBuilders { ... }
class KaleDefinition extends outer.Definition with KalePatternBuilders { ... }

val kale: Builders  = new KaleBuilders()

// the Pattern builders are still available but they are only meant to be used while constructing the definition
val setPatternDeclaredVeryEarly = kale.Application(Symbol("_ _"), Seq(1, 2)) 
// and eventually
val definition = kale.Definition(...)

// we could still allow the following, but it would be discouraged
definition.step(setPatternDeclaredVeryEarly)

// the users would need to/should use the Definition instead
val setDefinedThroughTheDefinition = definition.Application(Symbol("_ _"), Seq(1, 2)) 
definition.step(setDefinedThroughTheDefinition)
```

Yes, `Definition` could implement `OuterBuilders` as well (or the monolithic trait), but what would that actually mean? It would not be specific to this particular `Definition` so I would argue it should not be in its interface -- the users should do `kale.` instead.